File: server/src/lib/notion.js
// Notion: Aufgaben lesen/schreiben und Tages-/Wochenplaene als Seite anlegen.
import { config } from "../config.js";
import { request } from "./http.js";

const API = "https://api.notion.com/v1";
export const notionConfigured = () => Boolean(config.notion.token);

async function api(path, options = {}) {
  if (!notionConfigured()) throw new Error("Notion ist nicht konfiguriert (NOTION_TOKEN fehlt).");
  return request("notion", `${API}${path}`, {
    ...options,
    headers: {
      authorization: `Bearer ${config.notion.token}`,
      "Notion-Version": config.notion.version,
      "content-type": "application/json",
      ...(options.headers || {}),
    },
  });
}

// Notion-Datenbanken haben frei benannte Properties. Wir suchen die Spalten
// anhand ihres Typs, damit die Integration ohne exaktes Schema funktioniert.
async function tasksSchema() {
  const db = await api(`/databases/${config.notion.tasksDb}`);
  const props = db.properties || {};
  const byType = (type) => Object.entries(props).find(([, p]) => p.type === type)?.[0];
  const named = (re, type) =>
    Object.entries(props).find(([n, p]) => re.test(n) && (!type || p.type === type))?.[0];

  const status =
    named(/status|erledigt|done/i, "status") || byType("status") ||
    named(/erledigt|done/i, "checkbox") || byType("checkbox");

  return {
    title: byType("title"),
    status,
    statusType: status ? props[status].type : null,   // "status" oder "checkbox"
    date: named(/f(ä|ae)llig|due|datum|date/i, "date") || byType("date"),
    props,
  };
}

const plainTitle = (page, titleProp) =>
  (page.properties?.[titleProp]?.title || []).map((t) => t.plain_text).join("").trim();

export async function notionTasks({ includeDone = false } = {}) {
  if (!config.notion.tasksDb) return [];
  const s = await tasksSchema();
  const data = await api(`/databases/${config.notion.tasksDb}/query`, {
    method: "POST",
    body: JSON.stringify({ page_size: 100 }),
  });
  return (data.results || [])
    .map((p) => {
      const st = s.status ? p.properties[s.status] : null;
      const done = st?.type === "checkbox" ? st.checkbox : /done|fertig|erledigt|abgeschlossen/i.test(st?.status?.name || "");
      return {
        id: p.id,
        title: plainTitle(p, s.title) || "(ohne Titel)",
        done: Boolean(done),
        due: s.date ? p.properties[s.date]?.date?.start || null : null,
        url: p.url,
        source: "notion",
      };
    })
    .filter((t) => (includeDone ? true : !t.done));
}

export async function notionAddTask({ title, due }) {
  if (!config.notion.tasksDb) throw new Error("NOTION_TASKS_DB ist nicht gesetzt.");
  const s = await tasksSchema();
  const properties = { [s.title]: { title: [{ text: { content: title } }] } };
  if (due && s.date) properties[s.date] = { date: { start: due } };
  const p = await api("/pages", {
    method: "POST",
    body: JSON.stringify({ parent: { database_id: config.notion.tasksDb }, properties }),
  });
  return { id: p.id, title, due: due || null, url: p.url, source: "notion" };
}

export async function notionCompleteTask(pageId) {
  const s = await tasksSchema();
  if (!s.status) throw new Error("Keine Status- oder Checkbox-Spalte in der Notion-Datenbank gefunden.");

  let value;
  if (s.statusType === "checkbox") {
    value = { checkbox: true };
  } else {
    // Die Erledigt-Option heisst je nach Datenbank anders ("Done", "Erledigt",
    // "Fertig"). Deshalb aus den echten Optionen suchen statt "Done" zu raten.
    const options = s.props[s.status].status?.options || [];
    const groups = s.props[s.status].status?.groups || [];
    const completeGroup = groups.find((g) => /complete|done|fertig|abgeschlossen/i.test(g.name));
    const option =
      options.find((o) => /^(done|erledigt|fertig|abgeschlossen|completed)$/i.test(o.name)) ||
      (completeGroup && options.find((o) => completeGroup.option_ids?.includes(o.id))) ||
      options.at(-1);
    if (!option) throw new Error(`Die Spalte "${s.status}" hat keine Option für "erledigt".`);
    value = { status: { name: option.name } };
  }

  await api(`/pages/${pageId}`, { method: "PATCH", body: JSON.stringify({ properties: { [s.status]: value } }) });
  return true;
}

// Markdown-Untermenge -> Notion-Bloecke (Ueberschriften, Listen, To-dos, Absaetze).
function toBlocks(markdown) {
  const rich = (t) => [{ type: "text", text: { content: t.slice(0, 1900) } }];
  return markdown
    .split("\n")
    .map((line) => line.trimEnd())
    .filter((line) => line.length)
    .slice(0, 100)
    .map((line) => {
      let m;
      if ((m = line.match(/^###\s+(.*)/))) return { object: "block", type: "heading_3", heading_3: { rich_text: rich(m[1]) } };
      if ((m = line.match(/^##\s+(.*)/))) return { object: "block", type: "heading_2", heading_2: { rich_text: rich(m[1]) } };
      if ((m = line.match(/^#\s+(.*)/))) return { object: "block", type: "heading_1", heading_1: { rich_text: rich(m[1]) } };
      if ((m = line.match(/^[-*]\s+\[[ xX]\]\s+(.*)/)))
        return { object: "block", type: "to_do", to_do: { rich_text: rich(m[1]), checked: /\[[xX]\]/.test(line) } };
      if ((m = line.match(/^[-*]\s+(.*)/)))
        return { object: "block", type: "bulleted_list_item", bulleted_list_item: { rich_text: rich(m[1]) } };
      return { object: "block", type: "paragraph", paragraph: { rich_text: rich(line) } };
    });
}

export async function notionCreatePlanPage({ title, markdown }) {
  const parent = config.notion.planParent;
  if (!parent) throw new Error("NOTION_PLAN_PARENT ist nicht gesetzt.");
  // Der Parent darf eine Seite oder eine Datenbank sein - wir probieren beides.
  const asDatabase = {
    parent: { database_id: parent },
    properties: { title: { title: [{ text: { content: title } }] } },
    children: toBlocks(markdown),
  };
  const asPage = {
    parent: { page_id: parent },
    properties: { title: { title: [{ text: { content: title } }] } },
    children: toBlocks(markdown),
  };
  try {
    const p = await api("/pages", { method: "POST", body: JSON.stringify(asPage) });
    return { id: p.id, url: p.url, title };
  } catch {
    const p = await api("/pages", { method: "POST", body: JSON.stringify(asDatabase) });
    return { id: p.id, url: p.url, title };
  }
}

File: jarvis-local/jarvis/notion.py
"""Notion-Aufgaben ueber die offizielle REST-API.

Auth ist bewusst simpel gehalten: eine interne Integration mit einem Token.
Kein OAuth, kein Redirect. Du erstellst die Integration einmal, teilst die
Datenbank mit ihr, fertig.

Die Spalten werden nicht fest verdrahtet, sondern aus dem Schema erkannt.
Notions deutsche Aufgaben-Vorlage nennt sie "Aufgabenbezeichnung", "Status"
und "Faellig"; die englische "Name", "Status", "Due". Beides funktioniert,
ohne dass du etwas konfigurierst.
"""

from __future__ import annotations

import datetime as dt
import json
import re
import urllib.error
import urllib.request

API = "https://api.notion.com/v1"

# Kandidaten fuer die automatische Spaltenerkennung, beste zuerst
DATE_HINTS = ("fällig", "faellig", "due", "duedate", "datum", "date",
              "termin", "deadline", "abgabe")
DONE_HINTS = ("erledigt", "done", "fertig", "completed", "abgeschlossen")
STATUS_DONE_VALUES = ("done", "erledigt", "fertig", "abgeschlossen", "complete",
                      "completed", "archiviert")
PRIORITY_HINTS = ("priorität", "prioritaet", "priority", "wichtigkeit")


class NotionError(RuntimeError):
    pass


def _request(path: str, token: str, method: str = "GET",
             payload: dict | None = None, version: str = "2022-06-28",
             timeout: float = 20.0) -> dict:
    data = json.dumps(payload).encode("utf-8") if payload is not None else None
    request = urllib.request.Request(
        f"{API}{path}",
        data=data,
        method=method,
        headers={
            "Authorization": f"Bearer {token}",
            "Notion-Version": version,
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            return json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", "replace")
        try:
            message = json.loads(body).get("message", body)[:250]
        except json.JSONDecodeError:
            message = body[:250]
        if exc.code == 401:
            raise NotionError("Notion lehnt das Token ab. Neu erzeugen unter "
                              "notion.so/my-integrations") from exc
        if exc.code == 404:
            raise NotionError(
                "Datenbank nicht gefunden. Meist fehlt die Freigabe: in Notion "
                "die Datenbank oeffnen, oben rechts ··· > Verbindungen > "
                "deine Integration hinzufuegen."
            ) from exc
        if exc.code == 400 and "validation" in message.lower():
            raise NotionError(f"Notion weist die Anfrage ab: {message}") from exc
        raise NotionError(f"Notion Fehler {exc.code}: {message}") from exc
    except (urllib.error.URLError, TimeoutError, OSError) as exc:
        raise NotionError(f"Notion nicht erreichbar: {exc}") from exc


def normalize_id(raw: str) -> str:
    """Akzeptiert Datenbank-Link oder nackte ID und liefert die UUID."""
    raw = (raw or "").strip()
    # Aus einer URL die letzte 32-stellige Hex-Folge ziehen
    matches = re.findall(r"[0-9a-fA-F]{32}", raw.replace("-", ""))
    if matches:
        value = matches[-1].lower()
        return f"{value[0:8]}-{value[8:12]}-{value[12:16]}-{value[16:20]}-{value[20:32]}"
    return raw


def _plain_text(items) -> str:
    if not isinstance(items, list):
        return ""
    return "".join(part.get("plain_text", "") for part in items).strip()


class NotionTasks:
    def __init__(self, config):
        self.config = config
        self._schema: dict | None = None
        self.last_error: str | None = None

    @property
    def token(self) -> str:
        return self.config.get("notion.token", "") or ""

    @property
    def database_id(self) -> str:
        return normalize_id(self.config.get("notion.database_id", "") or "")

    @property
    def available(self) -> bool:
        return bool(self.config.get("notion.enabled", True)) and bool(
            self.token and self.database_id
        )

    # ------------------------------------------------------------------

    def schema(self, refresh: bool = False) -> dict:
        if self._schema is not None and not refresh:
            return self._schema
        data = _request(f"/databases/{self.database_id}", self.token,
                        version=self.config.get("notion.api_version", "2022-06-28"))
        properties = data.get("properties", {})
        self._schema = {
            "title": _plain_text(data.get("title")) or "Notion",
            "properties": {name: prop.get("type") for name, prop in properties.items()},
            "raw": properties,
        }
        return self._schema

    def _pick(self, wanted_types: tuple[str, ...], hints: tuple[str, ...]) -> str | None:
        """Spalte nach Typ und Namen erraten."""
        properties = self.schema()["properties"]
        candidates = [n for n, t in properties.items() if t in wanted_types]
        if not candidates:
            return None
        for hint in hints:
            for name in candidates:
                if hint in name.lower():
                    return name
        return candidates[0]

    def columns(self) -> dict:
        """Welche Spalte ist Titel, Datum, Erledigt-Kennzeichen?"""
        properties = self.schema()["properties"]
        title = next((n for n, t in properties.items() if t == "title"), None)
        return {
            "title": title,
            "date": self._pick(("date",), DATE_HINTS),
            "status": self._pick(("status",), ("status",)),
            "checkbox": self._pick(("checkbox",), DONE_HINTS),
            "priority": self._pick(("select", "status"), PRIORITY_HINTS),
            "database": self.schema()["title"],
        }

    # ------------------------------------------------------------------

    def _is_done(self, page: dict, columns: dict) -> bool:
        properties = page.get("properties", {})
        if columns.get("checkbox"):
            value = properties.get(columns["checkbox"], {})
            if value.get("type") == "checkbox":
                return bool(value.get("checkbox"))
        if columns.get("status"):
            value = properties.get(columns["status"], {}).get("status") or {}
            name = (value.get("name") or "").lower()
            group = (value.get("group") or {})
            group_name = (group.get("name") or "").lower() if isinstance(group, dict) else ""
            if name in STATUS_DONE_VALUES or group_name in ("complete", "done", "abgeschlossen"):
                return True
        return False

    def tasks(self, only_open: bool = True, limit: int = 60,
              horizon_days: int | None = None) -> list[dict]:
        """Aufgaben im selben Format wie die aus den Markdown-Notizen."""
        if not self.available:
            return []
        columns = self.columns()
        version = self.config.get("notion.api_version", "2022-06-28")

        payload: dict = {"page_size": min(limit, 100)}
        if columns.get("date"):
            payload["sorts"] = [{"property": columns["date"], "direction": "ascending"}]

        data = _request(f"/databases/{self.database_id}/query", self.token,
                        method="POST", payload=payload, version=version)

        today = dt.date.today()
        out: list[dict] = []
        for page in data.get("results", []):
            properties = page.get("properties", {})

            title = ""
            if columns.get("title"):
                title = _plain_text(properties.get(columns["title"], {}).get("title"))
            if not title:
                continue  # leere Zeilen gibt es in fast jeder Notion-Datenbank

            done = self._is_done(page, columns)
            if only_open and done:
                continue

            due = None
            if columns.get("date"):
                date_value = properties.get(columns["date"], {}).get("date") or {}
                start = date_value.get("start")
                if start:
                    try:
                        due = dt.date.fromisoformat(start[:10])
                    except ValueError:
                        due = None

            days_left = (due - today).days if due else None
            if horizon_days is not None and days_left is not None and days_left > horizon_days:
                continue

            status_name = ""
            if columns.get("status"):
                status_name = ((properties.get(columns["status"], {}).get("status") or {})
                               .get("name") or "")

            out.append({
                "text": title,
                "done": done,
                "source": f"Notion · {columns['database']}",
                "origin": "notion",
                "due": due.isoformat() if due else None,
                "days_left": days_left,
                "overdue": bool(due and not done and due < today),
                "tags": [status_name] if status_name else [],
                "url": page.get("url", ""),
                "status": status_name,
            })

        far = dt.date.max
        out.sort(key=lambda t: (
            dt.date.fromisoformat(t["due"]) if t["due"] else far, t["text"].lower()
        ))
        return out[:limit]

    def check(self) -> dict:
        """Fuer den Setup-Assistenten: Verbindung und erkannte Spalten pruefen."""
        columns = self.columns()
        sample = self.tasks(limit=5)
        return {
            "ok": True,
            "database": columns["database"],
            "columns": {k: v for k, v in columns.items() if k != "database"},
            "open_tasks": len(sample),
            "sample": [t["text"] for t in sample[:3]],
        }
